packages/shared/src/cachedFetch.ts

```diff
@@ -1,62 +1,106 @@
+import { createDeferred, Deferred } from "./async/createDeferred";
+import { timeoutAfter } from "./async/timeoutAfter";
 
-type CacheEntry = { json: any | null; status: number; statusText: string };
+type CacheEntry = {
+  json: any | null;
+  ok: boolean;
+  status: number;
+  statusText: string;
+};
 
-export const cache: Map<string, CacheEntry> = new Map();
+export const cache: Map<
+  string,
+  | {
+      deferred: Deferred<CacheEntry>;
+      entry: undefined;
+    }
+  | { deferred: undefined; entry: CacheEntry }
+> = new Map();
+
+type ShouldRetryValue = boolean | { after: number };
 
-// Note that this method should not be used for GraphQL queries because it caches responses by URL.
-// TODO [PRO-676] Move this into the "shared" package
+/** This method should not be used for GraphQL queries because it caches responses by URL. */
 export async function cachedFetch(
   url: string,
   init?: RequestInit,
   options: {
     baseDelay?: number;
     maxAttempts?: number;
-    shouldRetry?: (response: Response) => Promise<boolean>;
+    shouldRetry?: (
+      response: Response,
+      json: unknown,
+      defaultDelay: number
+    ) => Promise<ShouldRetryValue> | ShouldRetryValue;
   } = {}
 ): Promise<CacheEntry> {
   const { baseDelay = 1_000, maxAttempts = 3, shouldRetry: shouldRetryFn } = options;
 
-  let attempt = 0;
+  const cached = cache.get(url);
 
-  while (!cache.has(url)) {
-    attempt++;
+  if (cached) {
+    if (cached.entry) {
+      return cached.entry;
+    }
+    return cached.deferred.promise;
+  }
+
+  const deferred = createDeferred<CacheEntry>();
+  cache.set(url, { deferred, entry: undefined });
+
+  let attempt = 1;
+
+  while (true) {
+    // Retry with exponential backoff (e.g. 1s, 2s, 4s, ...) by default
+    let retryAfter = Math.pow(2, attempt) * baseDelay;
+    try {
+      const resp = await fetch(url, init);
+      const json = await resp.json().catch(() => null);
+
+      if (resp.ok) {
+        return storeCachedEntry(url, {
+          json,
+          ok: true,
+          status: resp.status,
+          statusText: resp.statusText,
+        });
+      }
 
-    const resp = await fetch(url, init);
-    if (resp.status === 200) {
-      const json = await resp.json();
-      cache.set(url, {
-        json,
-        status: resp.status,
-        statusText: resp.statusText,
-      });
-    } else if (attempt < maxAttempts) {
       if (shouldRetryFn) {
-        const shouldRetry = await shouldRetryFn(resp);
+        const shouldRetry = await shouldRetryFn(resp, json, retryAfter);
         if (!shouldRetry) {
-          cache.set(url, {
-            json: null,
+          return storeCachedEntry(url, {
+            json,
+            ok: false,
             status: resp.status,
             statusText: resp.statusText,
           });
-          break;
+        }
+        if (typeof shouldRetry === "object" && typeof shouldRetry.after === "number") {
+          retryAfter = shouldRetry.after;
```

Comment:
This seems like a bit of unnecessary complexity IMO, but maybe you had a reason to make this change? :)

Reply:
I need to control the delay in case I hit GitHub's rate limit. When hitting the secondary rate limit we need to wait like a minute or two (which is quite long... maybe I need to add some log about it too) and when hitting the primary rate limit the response contains the information about the time I need to wait before making the request again. 

I implemented this first using another optional callback but then I ended up checking the rate limit kind in both of those callbacks and it made sense to combine them into one (so `shouldRetry` could return the desired delay directly)

Reply:
Okay. That makes sense then.